### DISCUSSION TITLE
[Feature]: event Textbox from Resizing While Maintaining Position and Scaling in F*gma-like Behavior

### DISCUSSION THREAD

**Discussion by maxiannunziata** [fabricjs/fabric.js#10063] [Q&A]
### CheckList

- [X] I agree to follow this project's Code of Conduct
- [X] I have read and followed the [Contributing Guide](https://github.com/fabricjs/fabric.js/blob/master/CONTRIBUTING.md)
- [X] I have searched and referenced existing issues, feature requests and discussions
- [X] I am filing a **FEATURE** request.

### Description

Feature Request: Prevent Textbox from Resizing While Maintaining Position and Scaling in Figma-like Behavior

Example:
![Captura de pantalla 2024-08-13 a la(s) 7 34 09 a  m](https://github.com/user-attachments/assets/ec4e37e4-2fa1-4bc2-8c85-8064859cfaf3)
![Captura de pantalla 2024-08-13 a la(s) 7 34 02 a  m](https://github.com/user-attachments/assets/3fbc7303-3b59-44b1-aa0d-206b48c86105)

Description:

I would like to request a feature that allows the textbox to maintain its original size and position, scaling appropriately without losing its coordinates, similar to how Figma handles it.

Current Approach:

I have attempted to implement this behavior using the following code:

`updateZoom = () => {
  const zoom = this.canvas.getZoom();
  this.set({
    left: this.originalLeft / zoom,
    top: this.originalTop / zoom,
    scaleX: 1 / zoom,
    scaleY: 1 / zoom,
  });
  this.setCoords();
  this.canvas.requestRenderAll();
}
`

However, this approach is not entirely accurate. The textbox does not always maintain its exact position and scaling as expected, particularly when zoom levels change.

Preview: [https://imgflip.com/gif/905628](https://imgflip.com/gif/905628
)

Request:

It would be beneficial to have an out-of-the-box solution or feature that ensures the textbox does not resize or lose its position, while still scaling proportionally, similar to the behavior seen in design tools like Figma.

This would greatly enhance the user experience and precision when working with textboxes in dynamic canvases.

Thank you for considering this request!

### Current State

![Captura de pantalla 2024-08-13 a la(s) 7 47 08 a  m](https://github.com/user-attachments/assets/0e6864aa-c7a9-4e14-834a-ff98251a39fa)
![Captura de pantalla 2024-08-13 a la(s) 7 47 02 a  m](https://github.com/user-attachments/assets/198635d9-fb1f-4866-b7f2-4a088ccd414f)
![Captura de pantalla 2024-08-13 a la(s) 7 46 56 a  m](https://github.com/user-attachments/assets/328d48d7-c81e-46aa-adf7-713f2f316967)
![Captura de pantalla 2024-08-13 a la(s) 7 46 53 a  m](https://github.com/user-attachments/assets/455653b8-8698-432e-b6ee-3d55337ed3ba)

- Canvas Zooming: The canvas is set up with a zoom feature, allowing users to zoom in and out, which affects the scale and positioning of objects on the canvas.

- Textbox Scaling and Positioning: The SingleLineTextbox class adjusts its position (top, left) and font size dynamically in response to canvas zooming. However, achieving consistent behavior for text elements, particularly maintaining their size and position relative to the canvas during zooming, has proven challenging.

- Problem with Zoom: When the canvas zoom level changes, the font size, top, and left positions of the textboxes are not behaving as expected. The goal is to maintain the text's appearance (size, position) consistently regardless of the zoom level. The current implementation adjusts the font size and positions but has required fine-tuning and complex formulas to achieve the desired result.

- Non-linear Relationship: The relationship between the zoom level and the top position has been found to be non-linear, requiring a quadratic formula to describe it accurately.

### Additional Context

**Existing Solutions/Workarounds**

- Direct Adjustment of Font Size and Position: Initial approaches involved directly adjusting the font size and positions (top, left) based on the zoom level. This approach worked to some extent but introduced complexity, particularly when dealing with varying zoom levels and their non-linear impact.

- Quadratic Formula: A quadratic formula was derived to more accurately calculate the top position based on the zoom level. This formula was determined by fitting a curve to the data points of zoom levels and corresponding top positions. The formula is:

\[
\text{top} = -2411.64 \times (\text{zoom})^2 + 3199.48 \times \text{zoom} - 685.66
\]

- Custom Event Handling: Custom event handlers were introduced to listen for zoom changes and adjust the textbox properties accordingly. These handlers attempt to maintain the textbox's size and position consistent with the user's expectations during zooming.

- Iteration and Testing: Multiple iterations of testing and adjustments have been made to refine the behavior, particularly with the use of the quadratic formula to manage non-linear scaling issues.
-

**Comment by asturur** [comment]
The description of the issue and the behaviour is a bit confusing, i really do not understand what are you trying to achieve
Please remove Figma from your description and try to describe it by feature/functionality.

**Comment by drinkspiller** [comment]
@maxiannunziata , is [this](https://stackoverflow.com/a/79126627/13716260) what you're looking for?

**Comment by maxiannunziata** [comment]
Hi @drinkspiller thank you por the response!

I need [https://youtu.be/PR5ElnPnY2k](https://youtu.be/PR5ElnPnY2k) 

Thank you for your feedback!

To clarify, the feature I’m hoping to achieve is similar to how frames work in Figma when zooming in and out. Specifically, I would like a textbox to maintain a consistent position and size on the canvas without resizing or moving in an unintended way as the canvas zoom level changes.

What I’m Trying to Achieve:
In Figma, when you zoom in or out, frame titles remain anchored to their original position, appearing proportional and staying aligned without scaling or shifting unexpectedly. My goal is to replicate this behavior for textboxes on a zoomable canvas. This would mean that the textbox:

Keeps its initial position relative to the canvas (anchored coordinates).
Does not scale up or down visually, maintaining readability and consistency across zoom levels.
Current Challenges:
Currently, I'm adjusting position and scaling manually to try to achieve this effect, but it’s not resulting in the consistent, anchored positioning I’m looking for. The textbox still shifts or scales in ways that don’t look seamless with the zoom changes.

<img width="216" alt="Captura de pantalla 2024-11-13 a la(s) 7 53 34 a  m" src="https://github.com/user-attachments/assets/bb8005cd-95c2-4553-8917-f8a706378dc9">

Initial Behavior of the Title Frame
Description: In the first image, a title frame is positioned at the top of a main frame on a design canvas. This title frame acts as a label or title that identifies the content of the frame.

Behavior:

Anchored Position: The title frame is anchored to a specific corner of the main frame (typically the top-left corner). This means that regardless of zooming or canvas movement, the title maintains its relative position to the main frame.
Proportional Scaling: When zooming in or out, the title frame scales proportionally to maintain its readability and aesthetics. However, it does not resize disproportionately, which could affect its visibility.
Independence from Internal Content: The title frame is separated from the internal content of the main frame, allowing it to remain consistent without interfering with resizable or movable objects inside the frame.

-----

<img width="484" alt="Captura de pantalla 2024-11-13 a la(s) 7 53 29 a  m" src="https://github.com/user-attachments/assets/1aa1f333-8498-4f05-9b4b-0a8a2d17e234">

Description: The second image illustrates how the title frame behaves when zooming or resizing the main frame. It shows that the title continues to maintain its position and scale consistently.

Behavior:

Dynamic Adjustment: When zooming in or out, the title frame automatically readjusts to maintain its anchored position, ensuring that the title remains visible and properly aligned with the main frame.
Maintaining Proportions: The title frame keeps its original proportions, avoiding distortion or disproportionate scaling that could impact its appearance.
Adaptation to Resizing: When the main frame size changes, the title frame also adapts to keep its alignment and proportions, providing a seamless user experience.

**Comment by drinkspiller** [comment]
> Keeps its initial position relative to the canvas 

The [zoom example in the docs](https://fabricjs.com/docs/old-docs/fabric-intro-part-5/) explain how to to control the point the zoom is "anchored" to. TL;DR, [`canvas.zoomToPoint()`](https://fabricjs.com/api/classes/canvas/#zoomtopoint)

> When zooming in or out, the title frame scales proportionally to maintain its readability and aesthetics. However, it does not resize disproportionately

The [previous link ](https://stackoverflow.com/questions/20407546/scale-some-object-but-not-others-in-a-fabric-group/79126627#79126627) explains maintaining the appearance of not scaling. These may also help you achieve the desired result:

- [`lockScalingX`](https://fabricjs.com/api/classes/interactivefabricobject/#lockscalingx) and [`lockScalingY`](https://fabricjs.com/api/classes/interactivefabricobject/#lockscalingy)
- [`scaleToHeight()`](https://fabricjs.com/api/classes/interactivefabricobject/#scaletoheight)/[`scaleToWidth()`](https://fabricjs.com/api/classes/interactivefabricobject/#scaletowidth)
- https://fabricjs.com/api/classes/canvas/#uniformscaling
- https://fabricjs.com/api/classes/canvas/#uniscalekey

**Comment by asturur** [comment]
An object is bound to the canvas/document units, and the zoom is applied over as a blanket.
If you want to counter zoom effects you have to reverse it.

For example while zooming you have to go through this process.

My current viewport is a matrix X, so i take my position of the textbox and i transform it for the current matrix X, i get a sort of screen position. i keep it.

I change zoom and pan and now my viewport is X2, i have my screen position and i transform it by the inverse of X2 ( you have an utility to invert the matrix ).

Once you transform it by the inverse of the new viewport you can set it as the position of the textbox.

Same apply for scale.
